modules/core/src/test/java/org/apache/ignite/internal/processors/compute/ComputeGridMonitorMonitorTest.java

```diff
@@ -0,0 +1,431 @@
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(new NoopComputeTask(), null);
+
+        taskFut.get(getTestTimeout());
+
+        assertTrue(monitor.statusSnapshots.isEmpty());
+
+        assertEquals(3, monitor.statusDiffs.size());
+
+        checkStartTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkChangeJobNodes(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkFinishTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+    }
+
+    /**
+     * Checking get of diffs for the failed execution of the task.
+     */
+    @Test
+    public void failTaskTest() {
+        NoopComputeTask task = new NoopComputeTask() {
+            /** {@inheritDoc} */
+            @Override public Void reduce(List<ComputeJobResult> results) throws IgniteException {
+                throw new IgniteException("FAIL TASK");
+            }
+        };
+
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(task, null);
+
+        assertThrows(log, () -> taskFut.get(getTestTimeout()), IgniteException.class, null);
+
+        assertTrue(monitor.statusSnapshots.isEmpty());
+
+        assertEquals(3, monitor.statusDiffs.size());
+
+        checkStartTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkChangeJobNodes(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkFailTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+    }
+
+    @Test
+    public void changeAttributesTest() throws Exception {
+        ComputeFullWithWaitTask task = new ComputeFullWithWaitTask(getTestTimeout());
+
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(task, null);
+
+        task.doneOnMapFut.get(getTestTimeout());
+
+        taskFut.getTaskSession().setAttribute("test", "test");
+
+        task.waitOnJobExecFut.onDone();
+
+        taskFut.get(getTestTimeout());
+
+        assertTrue(monitor.statusSnapshots.isEmpty());
+
+        assertEquals(4, monitor.statusDiffs.size());
+
+        checkStartTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkChangeJobNodes(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkChangeAttributesNodes(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+        checkFinishTask(monitor.statusDiffs.poll(), taskFut.getTaskSession());
+    }
+
+    /**
+     * Checking the get of snapshots of task statuses.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void snapshotsTest() throws Exception {
+        ComputeFullWithWaitTask task = new ComputeFullWithWaitTask(getTestTimeout());
+
+        ComputeTaskFuture<Void> taskFut = CRD.compute().executeAsync(task, null);
+
+        task.doneOnMapFut.get(getTestTimeout());
+
+        ComputeGridMonitorImpl monitor1 = new ComputeGridMonitorImpl();
+
+        try {
+            CRD.context().task().listenStatusUpdates(monitor1);
+
+            task.waitOnJobExecFut.onDone();
```

Comment:
There should probably be another sleep after this line, otherwise test might become flaky. Or am I missing something?

Reply:
Deleted the unnecessary variable, now everything is fine